components/accounts/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+                    // coolbeans.png
+                    None
+                }
+                _ => None,
+            },
+            State::Authenticated => match via {
+                Event::AccountRestored => {
+                    log::info!("Ensuring device capabilities...");
+                    if let Err(err) = self
+                        .account
+                        .ensure_capabilities(&self.device_config.capabilities)
+                    {
+                        log::warn!("Failed to ensure device capabilities: {}", err);
+                    } else {
+                        log::info!("Successfully ensured device capabilities.");
+                    }
+                    Some(Event::FailedToAuthenticate)
+                }
+                Event::CompletedAuthFlow { code, state } => {
+                    log::info!("Completing OAuth flow...");
+                    if let Err(err) = self.account.complete_oauth_flow(code, state) {
+                        log::error!("Could not complete the OAuth flow: {}", err);
+                    }
+                    log::info!("Registering device...");
+                    if let Err(err) = self.account.initialize_device(
+                        &self.device_config.name,
+                        self.device_config.r#type,
+                        &self.device_config.capabilities,
+                    ) {
+                        log::warn!("Failed to register device: {}", err);
+                    } else {
+                        log::info!("Successfully registered device.");
+                    }
+                    None
+                }
+                Event::RecoveredFromAuthenticationProblem => {
+                    log::info!("Registering device...");
+                    if let Err(err) = self.account.initialize_device(
+                        &self.device_config.name,
+                        self.device_config.r#type,
+                        &self.device_config.capabilities,
+                    ) {
+                        log::warn!("Failed to register device: {}", err);
+                    } else {
+                        log::info!("Successfully registered device.");
+                    }
+                    None
+                }
+                _ => None,
+            },
+            State::AuthenticationProblem => match via {
+                Event::AuthenticationError => {
+                    log::info!("Hit authentication problem. Trying to recover.");
+                    self.account.clear_access_token_cache();
+                    if self.account.get_access_token(scopes::PROFILE).is_ok() {
+                        log::info!("Able to recover from an authentication problem.");
+                        return Some(Event::RecoveredFromAuthenticationProblem);
+                    }
+                    None
+                }
+                _ => None,
+            },
+        }
+    }
+
+    // Pums the state machine until it settles.
```

Comment:
There's a *lot* of machinery here for the state machine, and to be honest it feels a bit "forced" given that most of it is internal to this module. For example, we pump the state machine through potentially several states before it settles, but we don't provide any visibility into that outside of the FFI boundary, which makes we wonder if those intermediate states are really worth it.